### PR TITLE
ddl: refactor job args for alter sequence/auto-id-cache/shard-rowid/cache/ttl related ddl job.

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: Ref https://github.com/pingcap/tidb/issues/53930
Ref https://github.com/pingcap/tidb/issues/54436

Problem Summary:

### What changed and how does it work?
Refactor the job args for follow ddl types:
- ActionAlterSequence
- ActionModifyTableAutoIDCache
- ActionShardRowID
- ActionAlterCacheTable
- ActionAlterNoCacheTable
- ActionAlterTTLInfo
- ActionAlterTTLRemove

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/meta/model/job_args.go [hunk 1]
```diff
@@ -812,6 +812,123 @@ func GetRenameTablesArgs(job *Job) (*RenameTablesArgs, error) {
...
+			ShardRowIDBits: val,
+		}, nil
+	}
+
+	return getOrDecodeArgsV2[*ShardRowIDArgs](job)
+}
+
+// AlterTTLInfoArgs is the arguments for alter ttl info job.
+type AlterTTLInfoArgs struct {
+	TTLInfor           *TTLInfo `json:"ttl_infor,omitempty"`
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:906]:
```suggestion
	TTLInfo           *TTLInfo `json:"ttl_info,omitempty"`
```

File: pkg/ddl/executor.go [hunk 2]
```diff
@@ -3777,7 +3785,7 @@ func (e *executor) AlterTableRemoveTTL(ctx sessionctx.Context, ident ast.Ident)
 			CDCWriteSource: ctx.GetSessionVars().CDCWriteSource,
 			SQLMode:        ctx.GetSessionVars().SQLMode,
 		}
-		err = e.DoDDLJob(ctx, job)
+		err = e.doDDLJob2(ctx, job, nil)
```
Comment (D3Hunter) [inline pkg/ddl/executor.go:3788]: maybe add a EmptyArgs type for those types, so we don't need to consider nil

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: pkg/meta/model/job_args.go (hunk 1 above)
Comment (joccau) [reply pkg/meta/model/job_args.go]: Done

File: pkg/ddl/executor.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/ddl/executor.go:3804]: why resolve it?

but it's ok to handle it in other pr

File: pkg/ddl/executor.go [hunk 3]
```diff
@@ -6086,18 +6097,18 @@ func (e *executor) AlterTableCache(sctx sessionctx.Context, ti ast.Ident) (err e
...
 		TableID:        t.Meta().ID,
 		Type:           model.ActionAlterCacheTable,
 		BinlogInfo:     &model.HistoryInfo{},
-		Args:           []any{},
 		CDCWriteSource: sctx.GetSessionVars().CDCWriteSource,
 		SQLMode:        sctx.GetSessionVars().SQLMode,
 	}
 
-	return e.DoDDLJob(sctx, job)
+	return e.doDDLJob2(sctx, job, nil)
```
Comment (D3Hunter) [inline pkg/ddl/executor.go:6111]: doitto

File: pkg/meta/model/job_args.go [hunk 4]
```diff
@@ -55,6 +55,13 @@ type FinishedJobArgs interface {
 	fillFinishedJob(job *Job)
 }
 
+// EmptyArgs is the args for ddl job with nothing args.
+type EmptyArgs struct {
+}
+
+func (a *EmptyArgs) fillJob(_ *Job) {
+}
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:63]:
```suggestion
// EmptyArgs is the args for ddl job with nothing args.
type EmptyArgs struct {}

func (*EmptyArgs) fillJob(*Job) {}
```

File: pkg/ddl/table_test.go [hunk 5]
```diff
@@ -403,17 +403,17 @@ func testAlterCacheTable(
...
 		Type:       model.ActionAlterCacheTable,
 		BinlogInfo: &model.HistoryInfo{},
-		Args:       []any{},
 		InvolvingSchemaInfo: []model.InvolvingSchemaInfo{
 			{Database: newSchemaName.L, Table: tblInfo.Name.L},
 		},
 	}
 	ctx.SetValue(sessionctx.QueryString, "skip")
-	err := d.DoDDLJobWrapper(ctx, ddl.NewJobWrapper(job, true))
+	err := d.DoDDLJobWrapper(ctx, ddl.NewJobWrapperWithArgs(job, nil, true))
```
Comment (D3Hunter) [inline pkg/ddl/table_test.go:416]: doitto

below cases too

**Review by D3Hunter — APPROVED** [review]

**Comment by D3Hunter** [comment]
/hold

TestIndexMergePanic failed
```
[2024/09/24 08:33:36.091 +00:00] [ERROR] [misc.go:93] ["panic in the recoverable goroutine"] [r="failpoint panic: github.com/pingcap/tidb/pkg/executor/testIndexMergePanicPartialIndexWorker"] ["stack trace"="github.com/pingcap/tidb/pkg/util.WithRecovery.func1\n\tpkg/util/misc.go:95\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:920\ngithub.com/pingcap/failpoint.actPanic\n\texternal/com_github_pingcap_failpoint/terms.go:344\ngithub.com/pingcap/failpoint.(*term).do\n\texternal/com_github_pingcap_failpoint/terms.go:311\ngithub.com/pingcap/failpoint.(*terms).eval\n\texternal/com_github_pingcap_failpoint/terms.go:116\ngithub.com/pingcap/failpoint.(*Failpoint).Eval\n\texternal/com_github_pingcap_failpoint/failpoint.go:129\ngithub.com/pingcap/failpoint.(*Failpoints).Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:231\ngithub.com/pingcap/failpoint.Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:311\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1.1\n\tpkg/executor/index_merge_reader.go:362\ngithub.com/pingcap/tidb/pkg/util.WithRecovery\n\tpkg/util/misc.go:98\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1\n\tpkg/executor/index_merge_reader.go:360"] [stack="github.com/pingcap/tidb/pkg/util.WithRecovery.func1\n\tpkg/util/misc.go:93\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:920\ngithub.com/pingcap/failpoint.actPanic\n\texternal/com_github_pingcap_failpoint/terms.go:344\ngithub.com/pingcap/failpoint.(*term).do\n\texternal/com_github_pingcap_failpoint/terms.go:311\ngithub.com/pingcap/failpoint.(*terms).eval\n\texternal/com_github_pingcap_failpoint/terms.go:116\ngithub.com/pingcap/failpoint.(*Failpoint).Eval\n\texternal/com_github_pingcap_failpoint/failpoint.go:129\ngithub.com/pingcap/failpoint.(*Failpoints).Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:231\ngithub.com/pingcap/failpoint.Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:311\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1.1\n\tpkg/executor/index_merge_reader.go:362\ngithub.com/pingcap/tidb/pkg/util.WithRecovery\n\tpkg/util/misc.go:98\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1\n\tpkg/executor/index_merge_reader.go:360"]
```

**Comment by D3Hunter** [comment]
/unhold

seems not related to this pr

> /hold
> 
> TestIndexMergePanic failed, the panic is expected, but it timed out
> 
> ```
> [2024/09/24 08:33:36.091 +00:00] [ERROR] [misc.go:93] ["panic in the recoverable goroutine"] [r="failpoint panic: github.com/pingcap/tidb/pkg/executor/testIndexMergePanicPartialIndexWorker"] ["stack trace"="github.com/pingcap/tidb/pkg/util.WithRecovery.func1\n\tpkg/util/misc.go:95\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:920\ngithub.com/pingcap/failpoint.actPanic\n\texternal/com_github_pingcap_failpoint/terms.go:344\ngithub.com/pingcap/failpoint.(*term).do\n\texternal/com_github_pingcap_failpoint/terms.go:311\ngithub.com/pingcap/failpoint.(*terms).eval\n\texternal/com_github_pingcap_failpoint/terms.go:116\ngithub.com/pingcap/failpoint.(*Failpoint).Eval\n\texternal/com_github_pingcap_failpoint/failpoint.go:129\ngithub.com/pingcap/failpoint.(*Failpoints).Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:231\ngithub.com/pingcap/failpoint.Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:311\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1.1\n\tpkg/executor/index_merge_reader.go:362\ngithub.com/pingcap/tidb/pkg/util.WithRecovery\n\tpkg/util/misc.go:98\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1\n\tpkg/executor/index_merge_reader.go:360"] [stack="github.com/pingcap/tidb/pkg/util.WithRecovery.func1\n\tpkg/util/misc.go:93\nruntime.gopanic\n\tGOROOT/src/runtime/panic.go:920\ngithub.com/pingcap/failpoint.actPanic\n\texternal/com_github_pingcap_failpoint/terms.go:344\ngithub.com/pingcap/failpoint.(*term).do\n\texternal/com_github_pingcap_failpoint/terms.go:311\ngithub.com/pingcap/failpoint.(*terms).eval\n\texternal/com_github_pingcap_failpoint/terms.go:116\ngithub.com/pingcap/failpoint.(*Failpoint).Eval\n\texternal/com_github_pingcap_failpoint/failpoint.go:129\ngithub.com/pingcap/failpoint.(*Failpoints).Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:231\ngithub.com/pingcap/failpoint.Eval\n\texternal/com_github_pingcap_failpoint/failpoints.go:311\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1.1\n\tpkg/executor/index_merge_reader.go:362\ngithub.com/pingcap/tidb/pkg/util.WithRecovery\n\tpkg/util/misc.go:98\ngithub.com/pingcap/tidb/pkg/executor.(*IndexMergeReaderExecutor).startPartialIndexWorker.func1\n\tpkg/executor/index_merge_reader.go:360"]
> ```

File: pkg/meta/model/job_args.go [hunk 6]
```diff
@@ -55,6 +55,11 @@ type FinishedJobArgs interface {
 	fillFinishedJob(job *Job)
 }
 
+// EmptyArgs is the args for ddl job with nothing args.
```
Comment (joechenrh) [inline pkg/meta/model/job_args.go:58]:
```suggestion
// EmptyArgs is the args for ddl job with no args.
```

**Review by tangenta — APPROVED** [review]

**Pull request merged**